stdlib/asm/collections/smt.masm

```diff
@@ -301,3 +308,415 @@ export.get
+    swapw.2 loc_loadw.0 movdn.3 push.0 movup.3 push.0.0.0 loc_loadw.1 adv_push.4 loc_storew.0
+    # => [V_old, K, 0, 0, d, 0, R_new, N_old, ...]
+
+    # compute E = hash([K, V_old], domain=d)
+    # (10 cycle)
+    hperm dropw swapw dropw
+    # => [E, R_new, N_old, ...]
+
+    # make sure E and N_old are the same (14 cycles)
+    swapw swapw.2
+    repeat.4
+        dup.4 assert_eq
+    end
+    # => [E, R_new, ...]
+
+    # load the old value (which we saved previously) onto the stack (3 cycles)
+    loc_loadw.0
+    # => [V_old, R_new, ...]
+end
+
+#! Inserts a new leaf node at depth 16.
+#!
+#! Input: [V, K, R, ...];
+#! Output:[0, 0, 0, 0, R_new, ...]
+#!
+#! Where:
+#! - R is the initial root of the TSMT, and R_new is the new root of the TSMT.
+#! - K and V is the key-value pair for the leaf node to be inserted.
+#!
+#! This procedure succeeds only if:
+#! - Node to be replaced at depth 16 is a root of an empty subtree.
+#!
+#! Cycles: 73
+proc.insert_16
+    # extract 16-bit index from the key (8 cycles)
+    swapw dup exec.get_top_16_bits
+    # => [idx, K, V, R, ...]
+
+    # prepare the stack for computing leaf node value (6 cycles)
+    movdn.8 push.0.16.0.0 swapw.2
+    # => [V, K, 0, 0, 16, 0, idx, R, ...]
+
+    # compute leaf node value as N = hash([K, V], domain=16) (10 cycles)
+    hperm dropw swapw dropw
+    # => [N, idx, R, ...]
+
+    # prepare the stack for mtree_set operation (4 cycles)
+    swapw movup.8 movdn.4 push.16
+    # => [16, idx, R, N, ...]
+
+    # insert the node into the tree at depth 16; this also leaves the old value of the node on the
+    # stack (29 cycle)
+    mtree_set
+    # => [N_old, R_new, ...]
+
+    # verify that the old value of the node was a root of an empty subtree for depth 16 (12 cycles)
+    push.EMPTY_16_3 assert_eq
+    push.EMPTY_16_2 assert_eq
+    push.EMPTY_16_1 assert_eq
+    push.EMPTY_16_0 assert_eq
+
+    # put the return value onto the stack and return (4 cycles)
+    padw
+    # => [0, 0, 0, 0, R_new, ...]
+end
+
+#! Inserts a new leaf node at depths 32.
+#!
+#! Input: [V, K, R, ...];
+#! Output:[0, 0, 0, 0, R_new, ...]
+#!
+#! Where:
+#! - R is the initial root of the TSMT, and R_new is the new root of the TSMT.
+#! - K, V is the key-value pair for the leaf node to be inserted into the TSMT.
+#! - P is the internal node at depth 16 on the path to the new leaf node.
+#!
+#! This procedure consists of two high-level steps:
+#! - First, insert N = hash([K, V], domain=32) into a subtree with root P.
+#! - Then, insert the new root P into the TSMT with root R.
+#!
+#! We do this to minimize the number of hashes consumed by the procedure for Merkle path
+#! verification. Specifically, Merkle path verification will require exactly 64 hashes.
+#!
+#! This procedure succeeds only if:
+#! - Node at depth 16 is an internal node.
+#! - Node at depth 32 is a root of an empty subtree.
+#!
+#! Cycles: 183
+proc.insert_32.2
+    # load the value of P from the advice provider (5 cycles)
```

Comment:
Is P verified to be the correct node at the tier above anywhere?

Reply:
yes, we verify this later in the procedure when we do the first `mtree_set` and make sure that the old node value it leaves on the stack is the same as `P`.